### PR TITLE
#2921 Add color generator to shareholders chart

### PR DESCRIPTION
**Pull request by drakkein**

### CODE REVIEW

File: app/components/shared/charts/utils.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const hashCode = (str: string) => [...str].reduce((p, c) => c.charCodeAt(0) + ((p << 5) - p), 0);
```
Comment (AlexeyTuboltsev) [inline app/components/shared/charts/utils.ts]: This looks cool!
The problem with this is that it generates **random** colors, but our aim is to generate **contrasting** colors based on some initial value (like if we want 2 colors, we shouldn't get a white and black, but take some "neu_blue" as base color and get a contrasting yellowish color).  
As I see it, the algorithm would be to take base color (e.g. _#ffee25_), dissect it in _R_, _G_, _B_ channels (r: ff, g: ee, b: 25), divide the color space (0 to 255 or #00 to #ff) into number of segments (if we need 2 colors - add 255 to each channel, if we need 3 colors - divide span by 2 and add 127 to each channel for 2nd color and 255 for 3d color).  
By this you'll achieve that the colors are a) contrasting, b) are in a contrasting order - each color stands between contrasting ones. Also we wouldn't need predefined colors. They are anyway generated with some online color scheme generator using this algorithm :)).

But maybe your solution is sufficient, let's see. So this is non-blocking

File: app/lib/api/eto/EtoApi.interfaces.unsafe.tsx [hunk 2]
```diff
@@ -121,6 +121,15 @@ export const EtoKeyIndividualsType = YupTS.object({
 
 type TEtoKeyIndividualsType = YupTS.TypeOf<typeof EtoKeyIndividualsType>;
 
+const EtoLegalShareholderType = YupTS.object({
```
Comment (AlexeyTuboltsev) [inline app/lib/api/eto/EtoApi.interfaces.unsafe.tsx]: 👍

File: app/components/eto/public-view/LegalInformationWidget.tsx [hunk 3]
```diff
@@ -159,11 +163,15 @@ export const LegalInformationWidget: React.FunctionComponent<IProps> = ({
               data={{
                 datasets: [
                   {
-                    data: shareholdersData.map(d => d && d.shares),
-                    backgroundColor: shareholdersData.map((_, i: number) => CHART_COLORS[i]),
+                    data: shareholdersData!.map(d => d && d.shares),
```
Comment (AlexeyTuboltsev) [inline app/components/eto/public-view/LegalInformationWidget.tsx]: here and below there shouldn't be non-undefined assertions. `shareholdersData` is an array, empty or full, it never returns undefined. Please change the 'generateShareholders' signature to convince ts, don't use assertions

File: app/components/shared/charts/utils.ts (hunk 1 above)
Comment (drakkein) [reply app/components/shared/charts/utils.ts]: Let's summon @Moejoe90 to discuss this solution

Comment (Moejoe90) [reply app/components/shared/charts/utils.ts]: Let's test with this and see you're idea is good @AlexeyTuboltsev but I don't think we need to over-engineer this solution

Comment (AlexeyTuboltsev) [reply app/components/shared/charts/utils.ts]: hm. why over-engineering? I'd call it simply  engineering 😃

Comment (AlexeyTuboltsev) [reply app/components/shared/charts/utils.ts]: actually I forgot one step to make them in contrasting order

**Pull request merged**